File: src/main/kotlin/org/missionpinball/mpf/core/Device.kt
package org.missionpinball.mpf.core

import kotlinx.coroutines.CompletableDeferred

/**
 * Generic parent class for every hardware device in a pinball machine.
 *
 * One instance of a Device subclass is created for each device in the machine.
 */
abstract class Device(
    protected val machine: MachineController,
    val name: String
) {
    // Logging functionality via delegation
    private val logMixin: LogMixin = LogMixinImpl()

    protected val log get() = logMixin.log
    protected val debug get() = logMixin.debug

    /**
     * String of the config section name.
     * Must be overridden in subclasses.
     */
    abstract val configSection: String

    /**
     * String name of the collection.
     * Must be overridden in subclasses.
     */
    abstract val collection: String

    /**
     * String of the friendly name of the device class.
     * Must be overridden in subclasses.
     */
    abstract val classLabel: String

    /**
     * Can a config for this device be empty?
     */
    open val allowEmptyConfigs: Boolean = false

    /**
     * List of tags applied to this device.
     */
    val tags: MutableList<String> = mutableListOf()

    /**
     * Platform for this device (hardware interface).
     */
    var platform: Any? = null

    /**
     * User-friendly label for this device.
     */
    var label: String? = null

    /**
     * Validated dictionary of this device's settings.
     * Maps to the YAML-based config specified in the Config Spec.
     */
    var config: MutableMap<String, Any?> = mutableMapOf()

    /**
     * Compare two devices by name.
     */
    operator fun compareTo(other: Device): Int {
        return name.compareTo(other.name)
    }

    /**
     * Add a device to a running mode.
     *
     * @param mode Mode which loaded the device.
     * @throws IllegalStateException if device cannot be used in mode.
     */
    open suspend fun deviceAddedToMode(mode: Any) {
        throw IllegalStateException("Cannot use device $name in mode.")
    }

    /**
     * Called when device is added system-wide.
     */
    open suspend fun deviceAddedSystemWide() {
        // Initialize the device
        initialize()
    }

    /**
     * Get the config spec for this device.
     * Override in subclasses that have config specs.
     */
    open fun getConfigSpec(): Any? = null

    /**
     * Can be called in initialize() to load the platform section.
     *
     * @param platformSection Name of the platform section.
     */
    protected fun loadPlatformSection(platformSection: String) {
        val platformName = config["platform"] as? String
        // TODO: Implement when platform controller is available
        // platform = machine.getPlatformSections(platformSection, platformName)
    }

    /**
     * Return the config prepared for validation.
     *
     * @param deviceConfig Config of device.
     * @param isModeConfig Whether this device is loaded in a mode or system-wide.
     * @return Prepared config.
     */
    open fun prepareConfig(deviceConfig: Map<String, Any?>, isModeConfig: Boolean): Map<String, Any?> {
        return deviceConfig
    }

    /**
     * Return the parsed and validated config.
     *
     * @param deviceConfig Config of device.
     * @param isModeConfig Whether this device is loaded in a mode or system-wide.
     * @param debugPrefix Prefix to use when logging.
     * @return Validated config.
     */
    open fun validateAndParseConfig(
        deviceConfig: MutableMap<String, Any?>,
        isModeConfig: Boolean,
        debugPrefix: String? = null
    ): Map<String, Any?> {
        // TODO: Implement config validation when config_validator is available
        // machine.configValidator.validateConfig(configSection, deviceConfig, name, "device", prefix = debugPrefix)

        configureDeviceLogging(deviceConfig)

        return deviceConfig
    }

    /**
     * Configure logging for this device.
     */
    private fun configureDeviceLogging(deviceConfig: Map<String, Any?>) {
        var consoleLog = deviceConfig["console_log"] as? String ?: "basic"
        var fileLog = deviceConfig["file_log"] as? String ?: "basic"

        // If debug is enabled, set logging to full
        if (deviceConfig["debug"] as? Boolean == true) {
            consoleLog = "full"
            fileLog = "full"
        }

        logMixin.configureLogging(
            "$classLabel.$name",
            consoleLog,
            fileLog,
            classLabel
        )

        debugLog("Configuring device with settings: '$deviceConfig'")
    }

    /**
     * Load config for this device.
     *
     * @param deviceConfig Config for device.
     */
    open fun loadConfig(deviceConfig: Map<String, Any?>) {
        config = deviceConfig.toMutableMap()

        // Load tags
        @Suppress("UNCHECKED_CAST")
        (deviceConfig["tags"] as? List<String>)?.let {
            tags.clear()
            tags.addAll(it)
        }

        // Load label
        label = deviceConfig["label"] as? String
    }

    /**
     * Initialize device.
     * Override in subclasses to perform device-specific initialization.
     */
    protected open suspend fun initialize() {
        // Subclasses should override this
    }

    /**
     * Stop device.
     * Override in subclasses to perform cleanup when device is stopped.
     */
    open fun stopDevice() {
        // Subclasses can override this
    }

    override fun toString(): String {
        return "<$classLabel.$name>"
    }

    /**
     * Get config collection and config section.
     *
     * @return Pair with (collection, configSection).
     */
    fun getConfigInfo(): Pair<String, String> {
        require(collection.isNotEmpty() && configSection.isNotEmpty()) {
            "Implement collection and configSection in ${this::class.simpleName}"
        }
        return Pair(collection, configSection)
    }

    // Delegate logging methods to logMixin
    protected fun debugLog(msg: String, vararg args: Any?) {
        logMixin.debugLog(msg, *args)
    }

    protected fun infoLog(msg: String, vararg args: Any?) {
        logMixin.infoLog(msg, *args)
    }

    protected fun warningLog(msg: String, vararg args: Any?) {
        logMixin.warningLog(msg, *args)
    }

    protected fun errorLog(msg: String, vararg args: Any?) {
        logMixin.errorLog(msg, *args)
    }

    protected fun raiseConfigError(msg: String, errorNo: Int, loggerOverride: String? = null): Nothing {
        logMixin.raiseConfigError(msg, errorNo, loggerOverride)
    }

    companion object {
        /**
         * Device class initialization.
         * Called once per device class when devices are being created.
         * Override in subclasses that need class-level initialization.
         */
        @JvmStatic
        open fun deviceClassInit(machine: MachineController) {
            // Subclasses can override this
        }
    }
}

File: Assets/Scripts/Devices/Device.cs
using System;
using System.Collections.Generic;
using UnityEngine;

namespace MPF.Devices
{
    /// <summary>
    /// Base class for all MPF devices
    /// Converted from Python mpf/core/device.py
    /// </summary>
    public abstract class Device : MonoBehaviour
    {
        // String of the config section name (override in derived classes)
        public static string ConfigSection { get; protected set; }

        // String name of the collection (override in derived classes)
        public static string Collection { get; protected set; }

        // Friendly name of the device class
        public static string ClassLabel { get; protected set; }

        // Can a config for this device be empty?
        public static bool AllowEmptyConfigs { get; protected set; } = false;

        // Reference to machine controller
        protected MPF.Core.MachineController machine;

        // Device name
        public string DeviceName { get; protected set; }

        // Tags applied to this device
        public List<string> Tags { get; protected set; } = new List<string>();

        // Platform reference
        public object Platform { get; set; }

        // Label (friendly name)
        public string Label { get; set; }

        // Configuration dictionary
        public Dictionary<string, object> Config { get; protected set; } = new Dictionary<string, object>();

        // Debug mode
        public bool DebugMode { get; set; } = false;

        /// <summary>
        /// Initialize the device
        /// </summary>
        /// <param name="machineController">Reference to machine controller</param>
        /// <param name="name">Name of the device</param>
        public virtual void Initialize(MPF.Core.MachineController machineController, string name)
        {
            machine = machineController;
            DeviceName = name;
            gameObject.name = $"{GetType().Name}_{name}";

            if (DebugMode)
            {
                Debug.Log($"[Device] Initialized {GetType().Name}: {name}");
            }
        }

        /// <summary>
        /// Configure the device from a config dictionary
        /// </summary>
        public virtual void Configure(Dictionary<string, object> config)
        {
            Config = config ?? new Dictionary<string, object>();

            // Extract common config items
            if (Config.ContainsKey("label"))
            {
                Label = Config["label"].ToString();
            }

            if (Config.ContainsKey("tags"))
            {
                var tags = Config["tags"];
                if (tags is List<string> tagList)
                {
                    Tags = tagList;
                }
                else if (tags is string tagString)
                {
                    Tags = new List<string> { tagString };
                }
            }

            if (Config.ContainsKey("platform"))
            {
                // Platform loading happens in derived classes
            }
        }

        /// <summary>
        /// Validate and parse config
        /// Override in derived classes for specific validation
        /// </summary>
        public virtual Dictionary<string, object> ValidateAndParseConfig(Dictionary<string, object> config, bool isModeConfig)
        {
            // Base implementation just returns the config
            // Derived classes should override for validation
            return config;
        }

        /// <summary>
        /// Prepare config before validation
        /// </summary>
        public virtual Dictionary<string, object> PrepareConfig(Dictionary<string, object> config, bool isModeConfig)
        {
            // Base implementation - override in derived classes if needed
            return config;
        }

        /// <summary>
        /// Called when device is added to a mode
        /// </summary>
        public virtual void DeviceAddedToMode(object mode)
        {
            throw new InvalidOperationException($"Cannot use device {DeviceName} in mode");
        }

        /// <summary>
        /// Load platform section
        /// </summary>
        protected virtual void LoadPlatformSection(string platformSection)
        {
            // Platform loading - to be implemented when platforms are ready
            // this.Platform = machine.GetPlatformSections(platformSection, Config["platform"]);
        }

        /// <summary>
        /// Device-specific initialization (override in derived classes)
        /// </summary>
        protected virtual void DeviceInitialize()
        {
            // Override in derived classes
        }

        /// <summary>
        /// Stop the device
        /// </summary>
        public virtual void StopDevice()
        {
            if (DebugMode)
            {
                Debug.Log($"[Device] Stopping device: {DeviceName}");
            }
        }

        /// <summary>
        /// Compare devices by name
        /// </summary>
        public int CompareTo(Device other)
        {
            return string.Compare(DeviceName, other.DeviceName, StringComparison.Ordinal);
        }

        protected virtual void OnDestroy()
        {
            StopDevice();
        }
    }

    /// <summary>
    /// Base class for system-wide devices (not mode-specific)
    /// </summary>
    public abstract class SystemWideDevice : Device
    {
        // System-wide devices exist for the lifetime of the machine
    }

    /// <summary>
    /// Base class for mode-specific devices
    /// </summary>
    public abstract class ModeDevice : Device
    {
        // Mode devices are created/destroyed with modes
        public object Mode { get; set; }

        public override void DeviceAddedToMode(object mode)
        {
            Mode = mode;
        }
    }
}
